wordpress-mobile/WordPress-iOS
# [Tooling] Improvements on the AppLocalizedString linter

This is a follow-up on https://github.com/wordpress-mobile/WordPress-iOS/pull/18462 to address the remaining improvements I had in mind for the swift implementation back then:

## What was changed

 - Turn on optimizations when compiling the swift code (honestly doesn't change much, but 🤷 )
 - Remove debugging leftovers, fix typos in comments
 - Better strategy for parsing the `PBXObjects` based on their `isa`
 - Fix handling of `XCVersionGroup` and `PBXVariantGroup`
    - First by better understanding how to deal with the case of a `PBXBuildFile` pointing to such group was supposed to be handled in the first place…
    - … then by deciding to ignore those cases completely because they are only used for compiled resource "packages" like localized files (`*.strings`) and versioned files (`xcdatamodeld`), but we're only interested in source files (`*.swift`/`*.m`) anyway
 - Remove all force-unwraps and `fatalError`
    - Replace them with thrown errors
    - Provide better error messages and error reporting at top-level for any `.pbxproj` inconsistencies
    - More consistent reporting when encountering objects of known `isa` but with invalid data — as opposed to parse them as `UnknownPXBObject` at first and throw an unrelated error like `.orphanObject` only later
    - Protect against unknown values of some `PBXObject`'s fields, like an unknown `productType` or a `PBXBuildFile` pointing to a `productRef` (e.g. Swift Package dependency) instead of a `fileRef`

## To Test

 - Check that the project compiles in Xcode with a clean build
 - Check that the project compiles in Xcode with an incremental build
 - Check that the time taken by the Script Build phase in the build logs didn't jump to some crazy value and taking too long
 - Introduce a violation by replacing a random `AppLocalizedString` in code by `NSLocalizedString`, rebuild and validate the linter still detects it
 - Introduce an error in the `pbxproj` file to make it invalid, then check that the running the script from the terminal shows the expected error nicely. e.g.:
    - Replacing a `sourceTree = "<group>"` by `sourceTree = "wut"` to get a `DecodingError`
    - Changing a `PBXGroup` to have a `sourceTree = "<absolute>"` but removing its `path = …` value to get the `.incorrectAbsolutePath` error
    - Find the UUID of one of the parsed file, and remove it from the `children` list of its parent `PBXGroup` to get the `.orphanObject` error
    - Change an UUID randomly to get the `.objectNotFound` error
